tsinfer/inference.py

```diff
@@ -1102,36 +1165,59 @@ def match_samples_batch_partition(work_dir, partition_index):
         f"Matching partition {partition_index} with {partition_slice.start} to"
         f" {partition_slice.stop} of {len(sample_indexes)} samples"
     )
-    variant_data, ancestor_ts, matcher = load_variant_data_and_ancestors_ts(wd)
-    results = matcher.match_samples(
-        sample_indexes, sample_times, slice_=partition_slice
-    )
+    with provenance.TimingAndMemory() as timing:
+        variant_data, ancestor_ts, matcher = load_variant_data_and_ancestors_ts(wd)
+        results = matcher.match_samples(
+            sample_indexes, sample_times, slice_=partition_slice
+        )
     path = os.path.join(work_dir, f"partition_{partition_index}.pkl")
     logger.info(f"Dumping to {path}")
     with open(path, "wb") as f:
-        pickle.dump(results, f)
+        pickle.dump((timing.get_metrics(), results), f)
 
 
 def match_samples_batch_finalise(work_dir):
-    wd_path = os.path.join(work_dir, "wd.json")
-    wd = SampleBatchWorkDescriptor.load(wd_path)
-    variant_data, ancestor_ts, matcher = load_variant_data_and_ancestors_ts(wd)
-    results = []
-    for partition_index in range(wd.num_partitions):
-        path = os.path.join(work_dir, f"partition_{partition_index}.pkl")
-        with open(path, "rb") as f:
-            results.extend(pickle.load(f))
-    return match_samples(
-        variant_data,
-        ancestor_ts,
-        indexes=wd.indexes,
-        post_process=wd.post_process,
-        force_sample_times=wd.force_sample_times,
-        record_provenance=wd.record_provenance,
-        map_additional_sites=wd.map_additional_sites,
-        results=results,
-        **wd.common_params(),
-    )
+    with provenance.TimingAndMemory() as timing:
+        wd_path = os.path.join(work_dir, "wd.json")
+        wd = SampleBatchWorkDescriptor.load(wd_path)
+        variant_data, ancestor_ts, matcher = load_variant_data_and_ancestors_ts(wd)
+        results = []
+        timings = []
+        for partition_index in range(wd.num_partitions):
+            path = os.path.join(work_dir, f"partition_{partition_index}.pkl")
+            with open(path, "rb") as f:
+                part_timing, part_results = pickle.load(f)
+                results.extend(part_results)
+                timings.append(part_timing)
+        ts = match_samples(
+            variant_data,
+            ancestor_ts,
+            indexes=wd.indexes,
+            post_process=wd.post_process,
+            force_sample_times=wd.force_sample_times,
+            record_provenance=wd.record_provenance,
+            map_additional_sites=wd.map_additional_sites,
+            results=results,
+            **wd.common_params(),
+        )
+    # Rewrite the last provenance with the correct info
```

Comment:
Should be able to rewrite this using the dataclass, creating them with fromdict(), summing them totether and taking asdict of the result. That'll be simpler and less fragile.